### PR TITLE
Remove coupling of UI and game and player service

### PR DESCRIPTION
**Pull request by Sheikah45**
Game and Player service were putting updates to the game and player objects on the application thread in order to force a sense of synchronization to the updates. This caused an unnecessary load on the application thread and also blurred the lines between backend and frontend.

This pr separates out the ui based functions and the backend functions and ensures necessary synchronization on objects.

### CODE REVIEW

**Comment by Sheikah45** [comment]
Should be good to go

File: src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java [hunk 1]
```diff
@@ -146,9 +159,11 @@ ContextMenu getContextMenu() {
   }
 
   public void setChatUser(ChatChannelUser chatUser) {
+    if (this.chatUser != null) {
+      throw new IllegalStateException("Chat User already set");
+    }
```
Comment (micheljung) [inline src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java]:
```suggestion
    Assert.notNull(this.chatUser, "Chat user is already set");
```

File: src/main/java/com/faforever/client/game/Game.java [hunk 2]
```diff
@@ -253,30 +258,33 @@ public void setGameType(GameType gameType) {
...
 
   public MapProperty<String, String> simModsProperty() {
     return simMods;
   }
 
   /**
-   * Maps team names ("1", "2", ...) to a list of player names. <strong>Make sure to synchronize on the return
-   * value.</strong>
+   * Maps team names ("1", "2", ...) to a list of player names. <strong>Make sure to manually synchronize when iterating
+   * over * any of its collection views (e.g. values, entrySet, keySet).</strong> See {@link
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/Game.java]:
```suggestion
   * over any of its collection views (e.g. values, entrySet, keySet).</strong> See {@link
```

File: src/main/java/com/faforever/client/game/GameService.java [hunk 3]
```diff
@@ -270,14 +265,11 @@ public void changed(ObservableValue<? extends GameStatus> observable, GameStatus
...
-        boolean playerStillInGame = game.getTeams().entrySet().stream()
-            .flatMap(stringListEntry -> stringListEntry.getValue().stream())
-            .anyMatch(playerName -> playerName.equals(currentPlayer.getUsername()));
 
         /*
-          Check if player left the game while it was open, in this case we don't care any longer
+          Check if player still in the game
          */
-        if (newStatus == GameStatus.PLAYING && oldStatus == GameStatus.OPEN && !playerStillInGame) {
+        if (!(currentPlayer.getGame() == game) && game.getStatus() != GameStatus.CLOSED) {
```
Comment (micheljung) [inline src/main/java/com/faforever/client/game/GameService.java]:
```suggestion
        if (isCurrentPlayerInGame(game)) {
```

and add the `CLOSED`  check to the method.

**Review by micheljung — COMMENTED** [review]
As far as I can tell, solid work! Nice PR

File: src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java (hunk 1 above)
Comment (Sheikah45) [reply src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java]: Ah did not know that existed

**Pull request merged**